File: stack/BOJ/1874.cpp
#include <bits/stdc++.h>
using namespace std;

int main(void) {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n;
  cin >> n;
  stack<int> S;
  int num = 1; // 스택에 넣어야 할 숫자

  string res; // char 배열 대신 string 이용
  while (n--) {
    int t;
    cin >> t;
    while (num <= t) {
      S.push(num++);
      res += "+\n";
    }
    if (S.top() != t) {
      cout << "NO\n";
      return 0;
    }
    S.pop();
    res += "-\n";
  }
    cout << res;
}

/*
- char 배열 대신 string을 이용하는 것이 런타임 에러 발생을 줄일 수 있다.
- string에 문자를 추가할 때는 "+=" 이용
*/
// https://www.acmicpc.net/problem/1874

File: stack/BOJ/1874/1874.py
def main():   
    n = int(input())
    seq = []
    stack = []
    res = ''
    count = 1
    
    for i in range(n):
        seq.append(int(input()))
    
    for num in seq:
        if (stack):
            top = stack[len(stack) - 1]
            if (top == num):
                stack.pop()
                res += '-\n'
            elif (top < num):
                while (count <= num):
                    stack.append(count)
                    count += 1
                    res += '+\n'
                stack.pop()
                res += '-\n'
            else:
                res = 'NO\n'
                break
        else:
            while (count <= num):
                stack.append(count)
                count += 1
                res += '+\n'
            stack.pop()
            res += '-\n'
    print(res, end='')


if __name__ == "__main__":
    main()
